application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java

```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
+                typeRelation, savedDashboard.getUuidId(), EntityType.DASHBOARD
+        );
+        EntityRelation foundRelation = doGet(url, EntityRelation.class);
+        Assert.assertNotNull("Relation is not found!", foundRelation);
+        Assert.assertEquals("Found relation is not equals origin!", relation, foundRelation);
+
+        String dashboardIdStr = savedDashboard.getId().getId().toString();
+        doDelete("/api/dashboard/" + dashboardIdStr)
+                .andExpect(status().isBadRequest())
+                .andExpect(statusReason(containsString("The dashboard referenced by the device profiles cannot be deleted!")));
+
+        Dashboard afterErrorDeleteDashboard = doGet("/api/dashboard/" + dashboardIdStr,  Dashboard.class);
+        Assert.assertNotNull("Dashboard is not found!", afterErrorDeleteDashboard);
+        Assert.assertEquals("Dashboard after delete error is not equals origin!", savedDashboard, afterErrorDeleteDashboard);
+
+        EntityRelation afterErrorDeleteDashboardRelation = doGet(url, EntityRelation.class);
+        Assert.assertNotNull("Relation is not found!", foundRelation);
+        Assert.assertEquals("Relation after delete error Dashboard is not equals origin!", foundRelation, afterErrorDeleteDashboardRelation);
+
+        AlarmOperationResult afterErrorDeleteDashboardAlarmOperationResult = alarmService.createOrUpdateAlarm(savedAlarm);
+        Assert.assertEquals("List of propagatedEntities is not equal to number of created propagatedEntities!",
+                1, afterErrorDeleteDashboardAlarmOperationResult.getPropagatedEntitiesList().size());
+        Assert.assertEquals("DashboardId in propagatedEntities is not equal savedDashboardId!",
+                savedDashboard.getId(), afterErrorDeleteDashboardAlarmOperationResult.getPropagatedEntitiesList().get(0));
+
+        updatedDeviceProfile.setDefaultDashboardId(null);
+        doPost("/api/deviceProfile", updatedDeviceProfile)
+                .andExpect(status().isOk());
+
+        doDelete("/api/dashboard/" + dashboardIdStr)
+                .andExpect(status().isOk());
+        doGet("/api/dashboard/" + dashboardIdStr)
+                .andExpect(status().isNotFound())
+                .andExpect(statusReason(containsString(msgErrorNoFound("Dashboard", dashboardIdStr))));
+        doGet(url)
+                .andExpect(status().isNotFound())
+                .andExpect(statusReason(containsString(msgErrorNoFound("Dashboard", foundRelation.getTo().getId().toString()))));
+        AlarmOperationResult afterSuccessDeleteDashboardAlarmOperationResult = alarmService.createOrUpdateAlarm(savedAlarm);
+        Assert.assertEquals("List of propagatedEntities is not equal to number of created propagatedEntities!",
+                0, afterSuccessDeleteDashboardAlarmOperationResult.getPropagatedEntitiesList().size());
+    }
+
+    protected DeviceProfile createDeviceProfile() {
```

Comment:
the parent AbstractServiceTest already have method createDeviceProfile(). Please, use the existed method.
```
createDeviceProfile(tenantId, "my dev profile");
```